### DISCUSSION TITLE
Execution fails to generate pdf file

### DISCUSSION THREAD

**Discussion by jsbimra** [vercel/next.js#65884] [Help]
### Summary

POST

/api/generate-invoice

500
-74b8-463b-bd18-175080d1108f Version: $LATEST
[31m[PCR] Failed to create download dir: /home/sbx_user1051/.chromium-browser-snapshots[39m
[PCR] Puppeteer version: [35m21.11.0[39m
[PCR] Browser platform: [35mlinux[39m
[PCR] Chromium revision: [35m1302491[39m
[PCR] [33mNot found local chromium[39m
[PCR] Head request: https://storage.googleapis.com/chromium-browser-snapshots/Linux_x64/1302491/chrome-linux.zip - [32mtrue[39m (32ms)
[PCR] Head request: https://npmmirror.com/mirrors/chromium-browser-snapshots/Linux_x64/1302491/chrome-linux.zip - [31mfalse[39m (2175ms)
[PCR] Downloading from: https://storage.googleapis.com ...
[31m[PCR] Error: ENOENT: no such file or directory, mkdir '/home/sbx_user1051'[39m
[PCR] Downloading from: https://npmmirror.com/mirrors ...
[31m[PCR] Error: ENOENT: no such file or directory, mkdir '/home/sbx_user1051'[39m
[PCR] Retry Chromium downloading ... 
[PCR] Downloading from: https://storage.googleapis.com ...
[31m[PCR] Error: ENOENT: no such file or directory, mkdir '/home/sbx_user1051'[39m
[PCR] Downloading from: https://npmmirror.com/mirrors ...
[31m[PCR] Error: ENOENT: no such file or directory, mkdir '/home/sbx_user1051'[39m
[PCR] Retry Chromium downloading ... 
[PCR] Downloading from: https://storage.googleapis.com ...
[31m[PCR] Error: ENOENT: no such file or directory, mkdir '/home/sbx_user1051'[39m
[PCR] Downloading from: https://npmmirror.com/mirrors ...
[31m[PCR] Error: ENOENT: no such file or directory, mkdir '/home/sbx_user1051'[39m
[PCR] Retry Chromium downloading ... 
[PCR] Downloading from: https://storage.googleapis.com ...
[31m[PCR] Error: ENOENT: no such file or directory, mkdir '/home/sbx_user1051'[39m
[PCR] Downloading from: https://npmmirror.com/mirrors ...
[31m[PCR] Error: ENOENT: no such file or directory, mkdir '/home/sbx_user1051'[39m
[31m[PCR] ERROR: Failed to download Chromium after retry 3 times.[39m
[PCR] Chromium launchable: [31mfalse[39m
[31m[PCR] Failed to save stats: /home/sbx_user1051/.pcr-stats.json[39m
Error: An `executablePath` or `channel` must be specified for `puppeteer-core`
at assert (/var/task/node_modules/puppeteer-core/lib/cjs/puppeteer/util/assert.js:18:15)
at ChromeLauncher.computeLaunchArguments (/var/task/node_modules/puppeteer-core/lib/cjs/puppeteer/node/ChromeLauncher.js:102:36)
at ChromeLauncher.launch (/var/task/node_modules/puppeteer-core/lib/cjs/puppeteer/node/ProductLauncher.js:70:28)
at _default (/var/task/api/generate-invoice.js:50:25)
at Server.<anonymous> (/opt/rust/nodejs.js:1:10466)
at Server.<anonymous> (/opt/rust/nodejs.js:7:3490)
TypeError: Cannot read properties of undefined (reading 'newPage')
at _default (/var/task/api/generate-invoice.js:60:36)
at Server.<anonymous> (/opt/rust/nodejs.js:1:10466)
at Server.<anonymous> (/opt/rust/nodejs.js:7:3490)

Time
May 17 16:26:10.73 GMT+05:30
Status Code
500
Internal Server Error
Host
bizstore.doableyo.com
Request Path
/api/generate-invoice
Request ID
dzngp-1715943363114-f458d49b0ce8
Request User Agent
Mozilla/5.0 (Linux; Android 11; Redmi Note 8 Build/RKQ1.201004.002; wv) AppleWebKit/537.36 (KHTML, like Gecko) Version/4.0 Chrome/124.0.6367.179 Mobile Safari/537.36
Level
Error

Environment
production
Branch
main
Type
Serverless Function

Function
/api/generate-invoice
Location
Washington, D.C., USA (iad1)
Runtime
Node.js 18.x

Execution Duration / Limit
6.50s / 10s
Memory Used / Limit
111 MB / 1024 MB

**Comment by vieiralucas** [comment]
The two errors here are related:

1. `ENOENT: mkdir /home/sbx_user1051` — Vercel's serverless sandbox doesn't allow filesystem writes to that path. Puppeteer tries to create a user data directory on launch.

2. `An executablePath or channel must be specified` — `puppeteer-core` doesn't bundle Chromium, so you need to provide the binary path explicitly (usually via `@sparticuz/chromium`).

Even if you fix both, you'll likely hit the 50MB bundle limit or timeout issues. Puppeteer + Vercel serverless is a fundamentally fragile pairing.

A few alternatives that avoid running Chromium entirely:

- **[Gotenberg](https://gotenberg.dev)** — open source, self-hosted in Docker. Full Chromium rendering, you manage the infra.
- **[PDFShift](https://pdfshift.io)** — hosted API, free tier.
- **[Presa](https://presa.dev)** — hosted API I built, free tier (100 PDFs/month).

All take HTML and return a PDF via HTTP — works from any Vercel function without Chromium:

```typescript
import { PDF } from "presa-sdk";
const { buffer } = await PDF.generate({ html: invoiceHtml, format: "A4" });
```
